internal/credproviders/imds_provider.go

```diff
@@ -0,0 +1,104 @@
+// Copyright (C) MongoDB, Inc. 2023-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package credproviders
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"time"
+
+	"go.mongodb.org/mongo-driver/internal"
+	"go.mongodb.org/mongo-driver/internal/aws/credentials"
+)
+
+const (
+	// AzureProviderName provides a name of Azure provider
+	AzureProviderName = "AzureProvider"
+
+	azureURI = "http://169.254.169.254/metadata/identity/oauth2/token"
+)
+
+// An AzureProvider retrieves credentials from Azure IMDS.
+type AzureProvider struct {
+	httpClient   *http.Client
+	expiration   time.Time
+	expiryWindow time.Duration
+}
+
+// NewAzureProvider returns a pointer to an Azure credential provider.
+func NewAzureProvider(httpClient *http.Client, expiryWindow time.Duration) *AzureProvider {
+	return &AzureProvider{
+		httpClient:   httpClient,
+		expiration:   time.Time{},
+		expiryWindow: expiryWindow,
+	}
+}
+
+// RetrieveWithContext retrieves the keys from the Azure service.
+func (a *AzureProvider) RetrieveWithContext(ctx context.Context) (credentials.Value, error) {
+	v := credentials.Value{ProviderName: AzureProviderName}
+	req, err := http.NewRequest(http.MethodGet, azureURI, nil)
+	if err != nil {
+		return v, internal.WrapErrorf(err, "unable to retrieve Azure credentials")
```

Comment:
Optional: Now that the Go driver supports Go 1.13+, `internal.WrapErrorf` can be replaced with `fmt.Errorf` using the `%w` verb.

E.g.
```go
return v, fmt.Errorf("unable to retrieve Azure credentials: %w", err)
```

That suggestion applies to all usages of `internal.WrapErrorf`.